File: MD5.m
testStrings = {'siddharth', 'abinaya', 'sanjana'};
for i = 1:length(testStrings)
    s = testStrings{i};
    md5Hash = md5_digest(s);
    fprintf('MD5 Hash: %s for "%s"\n', md5Hash, s);
end

%%
function digest = md5_digest(message)
    % Constants
    m = 2 ^ 32;
    s = [-7, -12, -17, -22
         -5,  -9, -14, -20
         -4, -11, -16, -23
         -6, -10, -15, -21];
    t = floor(abs(sin(1:64)) .* m);
    digest = [hex2dec('67452301') ...
              hex2dec('EFCDAB89') ...
              hex2dec('98BADCFE') ...
              hex2dec('10325476')];

    % Convert message to bytes
    message = double(message);

    % the numel function is used to determine the number of elements in the variable message.
    bytelen = numel(message);

    % padding the message array, it appends 128 (binary representation of the bit 1 followed by seven 0 bits) to the message
    message = [message, 128, zeros(1, mod(55 - bytelen, 64))];

    % reshape the message array into a 2D array with 4 rows. 
    % The number of columns is equal to the total number of elements in message divided by 4.
    message = reshape(message, 4, numel(message) / 4);

    % combining each group of 4 bytes into a single 32-bit word
    message = message(1,:) + ...
              message(2,:) * 256 + ...
              message(3,:) * 65536 + ...
              message(4,:) * 16777216;
    
    % calculates the length of the original message in bits by multiplying the length in bytes by 8.
    bitlen = bytelen * 8;

    % appends the bit length of the original message to the end of the message array
    message = [message, mod(bitlen, m), mod(bitlen / m, m)];

    % Process each 512-bit block
    % iterate over the message array in chunks of 16 elements
    for k = 1:16:numel(message)

        % initialize the variables a, b, c, and d with the current values of the digest (initial condition are given above)
        a = digest(1); 
        b = digest(2); 
        c = digest(3); 
        d = digest(4);

        for i = (1:64)
            % convert the b , c, d to binary variable that is bv , cv, dv (bv -> binary variable)
            bv = dec2bin(b, 32) - '0';
            cv = dec2bin(c, 32) - '0';
            dv = dec2bin(d, 32) - '0';

            % list of operational logic is given here 
            if i <= 16
                f = (bv & cv) | (~bv & dv);
                ki = i - 1;
                sr = 1;

            elseif i <= 32
                f = (bv & dv) | (cv & ~dv);
                ki = mod(5 * i - 4, 16);
                sr = 2;

            elseif i <= 48
                f = xor(bv, xor(cv, dv));
                ki = mod(3 * i + 2, 16);
                sr = 3;

            else
                f = xor(cv, bv | ~dv);
                ki = mod(7 * i - 7, 16);
                sr = 4;
            end

            
            f = bin2dec(char(f + '0'));

            % calculate the circular shift number form that table 
            sc = mod(i - 1, 4) + 1;

            % calculates the sum of a, f, the current message word, and the current constant, modulo m.
            sum = mod(a + f + message(k + ki) + t(i), m);
            sum = dec2bin(sum, 32);

            % circular shift the sum by the number of bits given in the table
            sum = circshift(sum, [0, s(sr, sc)]);
            sum = bin2dec(sum);

            % the changing process happens here
            temp = d;
            d = c;
            c = b;
            b = mod(b + sum, m);
            a = temp;
        end

        % end of 512 bit block completeion (total 64 rounts has happened here)
        digest = mod(digest + [a, b, c, d], m);
    end

    % Convert hash to bytes and then to hexadecimal
    % division by powers of 256 is equivalent to a right shift of 8, 16, and 24 bits, respectively. 
    % This operation is used to split each 32-bit word in the digest into four 8-bit bytes
    % then we assemble it and return the final hash
    digest = [digest
              digest / 256
              digest / 65536
              digest / 16777216];
    digest = reshape(mod(floor(digest), 256), 1, numel(digest));
    digest = dec2hex(digest);
    digest = reshape(transpose(digest), 1, numel(digest));
    digest = lower(digest);
end

File: MD5.java
public class MD5 {

    // Constants for initializing variables in the MD5 algorithm
    private static final int INIT_A = 0x67452301;
    private static final int INIT_B = (int) 0xEFCDAB89L;
    private static final int INIT_C = (int) 0x98BADCFEL;
    private static final int INIT_D = 0x10325476;

    // Shift amounts for each round of the algorithm
    private static final int[] SHIFT_AMTS = {7, 12, 17, 22, 5, 9, 14, 20, 4, 11, 16, 23, 6, 10, 15, 21};

    // Lookup table with precomputed values
    private static final int[] TABLE_T = new int[64];

    // Static block to initialize the lookup table
    static {
        for (int i = 0; i < 64; i++)
            // Initialize the lookup table with values based on the sine function
            TABLE_T[i] = (int) (long) ((1L << 32) * Math.abs(Math.sin(i + 1)));
    }

    // Method to compute MD5 hash of a given message
    public static byte[] computeMD5(byte[] message) {

        // Calculate the length of the message in bytes
        int messageLenBytes = message.length;

        // Calculate the number of 64-byte blocks required, including padding
        int numBlocks = ((messageLenBytes + 8) >>> 6) + 1;

        // Calculate the total length of the padded message
        int totalLen = numBlocks << 6;

        // Create an array for padding bytes
        byte[] paddingBytes = new byte[totalLen - messageLenBytes];
        paddingBytes[0] = (byte) 0x80; // Set the first padding byte to 0x80

        // Calculate the length of the message in bits and store it in the padding
        long messageLenBits = (long) messageLenBytes << 3;
        for (int i = 0; i < 8; i++) {
            paddingBytes[paddingBytes.length - 8 + i] = (byte) messageLenBits;
            messageLenBits >>>= 8;
        }

        // Initialize variables for the MD5 algorithm
        int a = INIT_A;
        int b = INIT_B;
        int c = INIT_C;
        int d = INIT_D;
        int[] buffer = new int[16];

        // Process each 64-byte block
        for (int i = 0; i < numBlocks; i++) {

            // Calculate the index for the current block
            int index = i << 6;

            // Fill the buffer with data from the message or padding
            for (int j = 0; j < 64; j++, index++)
                
                buffer[j >>> 2] = ((int) ((index < messageLenBytes) ? message[index]
                        : paddingBytes[index - messageLenBytes]) << 24)
                        | (buffer[j >>> 2] >>> 8);

            // Save the original values of variables
            int originalA = a;
            int originalB = b;
            int originalC = c;
            int originalD = d;

            // Process each of the 64 rounds
            for (int j = 0; j < 64; j++) {
                int div16 = j >>> 4;
                int f = 0;
                int bufferIndex = j;

                // Determine the function for the current round
                switch (div16) {
                    case 0:
                        f = (b & c) | (~b & d);
                        break;
                    case 1:
                        f = (b & d) | (c & ~d);
                        bufferIndex = (bufferIndex * 5 + 1) & 0x0F;
                        break;
                    case 2:
                        f = b ^ c ^ d;
                        bufferIndex = (bufferIndex * 3 + 5) & 0x0F;
                        break;
                    case 3:
                        f = c ^ (b | ~d);
                        bufferIndex = (bufferIndex * 7) & 0x0F;
                        break;
                }

                // Update temporary variables using bitwise operations
                int temp = b + Integer.rotateLeft(a + f + buffer[bufferIndex] + TABLE_T[j],
                        SHIFT_AMTS[(div16 << 2) | (j & 3)]);
                a = d;
                d = c;
                c = b;
                b = temp;
            }

            // Update main variables
            a += originalA;
            b += originalB;
            c += originalC;
            d += originalD;
        }

        // Concatenate the four 32-bit variables to produce the MD5 hash
        byte[] md5 = new byte[16];
        int count = 0;
        for (int i = 0; i < 4; i++) {
            int n = (i == 0) ? a : ((i == 1) ? b : ((i == 2) ? c : d));
            for (int j = 0; j < 4; j++) {
                md5[count++] = (byte) n;
                n >>>= 8;
            }
        }

        return md5;
    }

    // Method to convert a byte array to a hexadecimal string
    public static String toHexString(byte[] b) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < b.length; i++) {
            sb.append(String.format("%02X", b[i] & 0xFF));
        }
        return sb.toString();
    }

    // Main method for testing the MD5 implementation
    public static void main(String[] args) {
        String[] testStrings = {"siddharth"};
        for (String s : testStrings)
            System.out.println(toHexString(computeMD5(s.getBytes())) + " <== \"" + s + "\"");
    }
}
